core/src/main/scala/stryker4s/config/ConfigReader.scala

```diff
@@ -4,45 +4,152 @@ import java.io.FileNotFoundException
+import pureconfig.{ConfigSource, Derivation, ConfigReader => PureConfigReader}
 import stryker4s.config.implicits.ConfigReaderImplicits
+import cats.syntax.either._
 import pureconfig.generic.auto._
 
 object ConfigReader extends ConfigReaderImplicits with Logging {
 
   val defaultConfigFileLocation: File = File.currentWorkingDirectory / "stryker4s.conf"
 
+  private val configDocUrl: String =
+    "https://github.com/stryker-mutator/stryker4s/blob/master/docs/CONFIGURATION.md"
+
   /** Read config from stryker4s.conf. Or use the default Config if no config file is found.
     */
-  def readConfig(confFile: File = defaultConfigFileLocation): Config = readConfigOfType[Config](confFile) fold (
-    tryRecoverFromFailures,
-    logAndReturn(info("Using stryker4s.conf in the current working directory"))
-  )
+  def readConfig(confFile: File = defaultConfigFileLocation): Config = {
+    implicit val hint: ProductHint[Config] = ProductHint[Config](allowUnknownKeys = false)
+
+    Reader[Config](confFile)
+      .recoverWithDerivation(Failure.onUnknownKey)
+      .recoverWith(Failure.onFileNotFound.andThen(_.asRight[ConfigReaderFailures]))
+      .config
+  }
 
   def readConfigOfType[T](
       confFile: File = defaultConfigFileLocation
   )(implicit derivation: Derivation[PureConfigReader[T]]): Either[ConfigReaderFailures, T] =
-    ConfigSource.file(confFile.path).at("stryker4s").load[T]
-
-  private def tryRecoverFromFailures[T](failures: ConfigReaderFailures): Config = failures match {
-    case ConfigReaderFailures(CannotReadFile(fileName, Some(_: FileNotFoundException)), _) =>
-      warn(s"Could not find config file $fileName")
-      warn("Using default config instead...")
-      // FIXME: sbt has its own (older) dependency on Typesafe config, which causes an error with Pureconfig when running the sbt plugin
-      //  If that's fixed we can add this again
-      //  https://github.com/stryker-mutator/stryker4s/issues/116
-      // info("Config used: " + defaultConf.toHoconString)
-
-      Config()
-    case _ =>
+    Reader[T](confFile).tryRead
+
+  /**
+    * A configuration on how to attempt to read a config. The reason for its existence is to
+    * provide a convenient way to attempt to read a config with various [[Derivation]]s,
+    * depending on the [[ConfigReaderFailures]] that was returned from the last attempt.
+    * In addition to that, some logging is also enforced during the reading process.
+    * If not for those points, simply using the returned [[Either]] of the [[PureConfigReader]]
+    * would be sufficient as well, since the exposed API here is basically a subset of [[Either]]s.
+    * @param file the [[File]] from which the config is to be read.
+    * @param derivation the [[Derivation]] that is used to configure the [[PureConfigReader]].
+    * @tparam T the type of the config that is to be read.
+    */
+  private class Reader[T] private (file: File)(implicit derivation: Derivation[PureConfigReader[T]]) {
```

Comment:
I like this construction of a builder-like reader with composable behaviour a lot! However, I feel like the mutability can be avoided. The functions could be in the constructor, with the empty PF as defaults. Each function that returns a `Reader` could return a new instance of a `Reader` with the new functions composed.

Lastly, this class is getting quite big (in part due to the comments 👍). Could you move it to a separate file and extract any config-reader specific stuff from it (like logging), just to separate the concerns and make testing easier?

Reply:
I will remove the mutable instance. I originally placed it here since we are in very controlled, safe space and thus it felt ok to avoid to cost of instantiating a new object for every api-call but since reading the config is not on any hot-path whatsoever, I agree that this is unnecessary.

I can also move the class, but I'm not sure about pulling the logging out of it. One of the reason for it's existence was to force some level of common logging when reading configs of different types. 
Taking another look at the logging though made me realise that the info that is logged on a successful read is wrong. The current text is only correct for the default file-path. 
How about I change the logging such that the path of the config that was read is taken from the input to the class?

Reply:
I am not sure anymore about moving the class. What should be its scope? It's only functionality is to configure the reading from config. That's exactly what `ConfigReader` does. If I remove it from the `ConfigReader`, that object will only contain a call to that class. 
Isn't it good that we are testing the `ConfigReader` and not its implementation details?

Reply:
> If I remove it from the ConfigReader, that object will only contain a call to that class.
Isn't it good that we are testing the ConfigReader and not its implementation details?

Taking another looking at it and reading your comment, I agree! It makes perfect sense for it to be an inner class 👍